File: core/src/main/java/com/vzome/core/commands/ZomicVirtualMachine.java

package com.vzome.core.commands;

import com.vzome.core.algebra.AlgebraicNumber;
import com.vzome.core.algebra.AlgebraicVector;
import com.vzome.core.construction.AnchoredSegment;
import com.vzome.core.construction.Construction;
import com.vzome.core.construction.ConstructionChanges;
import com.vzome.core.construction.Point;
import com.vzome.core.construction.Segment;
import com.vzome.core.construction.SegmentEndPoint;
import com.vzome.core.math.symmetry.Axis;
import com.vzome.core.math.symmetry.Symmetry;
import com.vzome.core.render.AbstractZomicEventHandler;


public class ZomicVirtualMachine extends AbstractZomicEventHandler
{
    private Point mLocation;
    
    private final ConstructionChanges mEffects;

    public AlgebraicVector getLocation()
    {
        return mLocation .getLocation();
    }
    
    /**
     * @return
     */
    public Construction getLastPoint()
    {
        return mLocation;
    }

    @Override
    public void step( Axis axis, AlgebraicNumber length )
    {
        axis = mOrientation .permute( axis, mHandedNess );
        length = length .times( mScale );
        
        Segment segment = new AnchoredSegment( axis, length, mLocation );
        Point pt2 = new SegmentEndPoint( segment );
        
        if ( mAction != JUST_MOVE ) {
            mEffects .constructionAdded( mLocation );
            mEffects .constructionAdded( segment );
            mEffects .constructionAdded( pt2 );
//            mEffects .constructionHidden( segment );
//            mEffects .constructionHidden( pt2 );  // TODO THIS IS A BUG!
        }
        mLocation = pt2;
    }
    
    public ZomicVirtualMachine( Point start, ConstructionChanges effects, Symmetry symm )
    {
        super( symm );
        mLocation = start;
        mEffects = effects;
    }

    @Override
    protected AbstractZomicEventHandler copyLocation()
    {
        return new ZomicVirtualMachine( mLocation, mEffects, mSymmetry );
    }

    @Override
    protected void restoreLocation( AbstractZomicEventHandler changed )
    {
        mLocation = ((ZomicVirtualMachine) changed) .mLocation;
    }
    
}

File: online/src/worker/legacy/from-java/com/vzome/core/commands/ZomicVirtualMachine.ts
import { AlgebraicNumber } from "../algebra/AlgebraicNumber.js";
import { AlgebraicVector } from "../algebra/AlgebraicVector.js";
import { AnchoredSegment } from "../construction/AnchoredSegment.js";
import { Construction } from "../construction/Construction.js";
import { ConstructionChanges } from "../construction/ConstructionChanges.js";
import { Point } from "../construction/Point.js";
import { Segment } from "../construction/Segment.js";
import { SegmentEndPoint } from "../construction/SegmentEndPoint.js";
import { Axis } from "../math/symmetry/Axis.js";
import { Symmetry } from "../math/symmetry/Symmetry.js";
import { AbstractZomicEventHandler } from "../render/AbstractZomicEventHandler.js";
import { ZomicEventHandler } from "../render/ZomicEventHandler.js";

export class ZomicVirtualMachine extends AbstractZomicEventHandler {
    /*private*/ mLocation: Point;

    /*private*/ mEffects: ConstructionChanges;

    public getLocation(): AlgebraicVector {
        return this.mLocation.getLocation();
    }

    /**
     * @return
     * @return {Construction}
     */
    public getLastPoint(): Construction {
        return this.mLocation;
    }

    /**
     * 
     * @param {Axis} axis
     * @param {*} length
     */
    public step(axis: Axis, length: AlgebraicNumber) {
        axis = this.mOrientation.permute(axis, this.mHandedNess);
        length = length['times$com_vzome_core_algebra_AlgebraicNumber'](this.mScale);
        const segment: Segment = new AnchoredSegment(axis, length, this.mLocation);
        const pt2: Point = new SegmentEndPoint(segment);
        if (this.mAction !== ZomicEventHandler.JUST_MOVE){
            this.mEffects['constructionAdded$com_vzome_core_construction_Construction'](this.mLocation);
            this.mEffects['constructionAdded$com_vzome_core_construction_Construction'](segment);
            this.mEffects['constructionAdded$com_vzome_core_construction_Construction'](pt2);
        }
        this.mLocation = pt2;
    }

    public constructor(start: Point, effects: ConstructionChanges, symm: Symmetry) {
        super(symm);
        if (this.mLocation === undefined) { this.mLocation = null; }
        if (this.mEffects === undefined) { this.mEffects = null; }
        this.mLocation = start;
        this.mEffects = effects;
    }

    /**
     * 
     * @return {AbstractZomicEventHandler}
     */
    copyLocation(): AbstractZomicEventHandler {
        return new ZomicVirtualMachine(this.mLocation, this.mEffects, this.mSymmetry);
    }

    /**
     * 
     * @param {AbstractZomicEventHandler} changed
     */
    restoreLocation(changed: AbstractZomicEventHandler) {
        this.mLocation = (<ZomicVirtualMachine>changed).mLocation;
    }
}
ZomicVirtualMachine["__class"] = "com.vzome.core.commands.ZomicVirtualMachine";
ZomicVirtualMachine["__interfaces"] = ["com.vzome.core.render.ZomicEventHandler"];
